File: Advent of Code/Challenge Solutions/Year 2025/ChallengeSolution04.cs
// Task: https://adventofcode.com/2025/day/4

using Advent_of_Code.Shared;
using Advent_of_Code.Shared.Utilities;

namespace Advent_of_Code.Challenge_Solutions.Year_2025;

public class ChallengeSolution04(IConsole console, ISolutionReader<ChallengeSolution04> reader)
    : ChallengeSolution<ChallengeSolution04>(console, reader)
{
    private const char Empty = '.';
    private const char PaperRoll = '@';
    private const int MaxPaperRolls = 4;

    public override void SolveFirstPart()
    {
        var grid = ReadGrid();

        var reachablePaperRolls = 0;

        for (var row = 0; row < grid.Length; row++)
        {
            for (var col = 0; col < grid.Length; col++)
            {
                if (IsAccessiblePaperRollPosition(grid, row, col))
                {
                    reachablePaperRolls++;
                }
            }
        }

        Console.WriteLine($"Reachable paper rolls: {reachablePaperRolls}");
    }

    public override void SolveSecondPart()
    {
        var grid = ReadGrid();

        var removedPaperRolls = 0;

        var noMoreRollsAccessible = false;
        while (!noMoreRollsAccessible)
        {
            noMoreRollsAccessible = true;
            List<(int, int)> rollsToRemove = [];

            for (var row = 0; row < grid.Length; row++)
            {
                for (var col = 0; col < grid.Length; col++)
                {
                    if (IsAccessiblePaperRollPosition(grid, row, col))
                    {
                        noMoreRollsAccessible = false;
                        rollsToRemove.Add((row, col));
                    }
                }
            }

            foreach (var (row, col) in rollsToRemove)
            {
                grid[row][col] = Empty;
            }

            removedPaperRolls += rollsToRemove.Count;
        }

        Console.WriteLine($"Removed paper rolls: {removedPaperRolls}");
    }

    private static bool IsAccessiblePaperRollPosition(char[][] grid, int row, int col)
    {
        if (grid[row][col] == Empty)
        {
            return false;
        }

        return GetNeighbours(grid, row, col)
            .Count(neighbour => neighbour == PaperRoll) < MaxPaperRolls;
    }

    private static IEnumerable<char> GetNeighbours(char[][] grid, int row, int col)
    {
        List<(int, int)> potentialNeighbouringPositions =
            [
                (row - 1, col),
                (row + 1, col),
                (row, col - 1),
                (row, col + 1),
                (row - 1, col - 1),
                (row - 1, col + 1),
                (row + 1, col - 1),
                (row + 1, col + 1),
            ];

        foreach (var (i, j) in potentialNeighbouringPositions)
        {
            if (i < 0 || i >= grid.Length
                || j < 0 || j >= grid[i].Length)
            {
                continue;
            }

            yield return grid[i][j];
        }
    }

    private char[][] ReadGrid() => Reader
        .ReadLines()
        .Select(line => line.ToCharArray())
        .ToArray();
}

File: Advent of Code.FSharp/Challenge Solutions/Year 2025/ChallengeSolution04.fs
// Task: https://adventofcode.com/2025/day/4

namespace Advent_of_Code.FSharp.Challenge_Solutions.Year_2025

open Advent_of_Code.Shared
open Advent_of_Code.Shared.Utilities
open Advent_of_Code.FSharp.Utilities

module ChallengeSolution04 =

    [<Literal>]
    let maxAdjacentPaperRolls = 4
    
    type Floor =
        | PaperRoll
        | Empty
        
    type Grid = Floor[,]
    
    type PaperRollLocation = int * int
    
    type Solution(getInput: unit -> string list, writeLine: string -> unit) =
        
        let readFloorGrid() : Grid =
            getInput()
            |> Seq.map
                ^<| Seq.map (function
                        | '@' -> PaperRoll
                        | _   -> Empty
                    )
            |> array2D
            
        let getAllPaperRolls (grid: Grid) : PaperRollLocation list =
            grid
            |> Array2D.mapi (fun row col value ->
                             match value with
                             | PaperRoll -> Some (row, col)
                             | _ -> None)
            |> Seq.cast<PaperRollLocation option>
            |> Seq.choose id
            |> List.ofSeq
            
        let getNeighbouringPaperRolls (grid: Grid) (row: int, col: int) =
            [
                (row - 1, col)
                (row + 1, col)
                (row, col - 1)
                (row, col + 1)
                (row - 1, col - 1)
                (row - 1, col + 1)
                (row + 1, col - 1)
                (row + 1, col + 1)
            ]
            |> List.filter (fun (r, c) -> r >= 0 && r < (grid |> Array2D.length1)
                                           && c >= 0 && c < (grid |> Array2D.length2)
                                           && grid[r, c] = PaperRoll)
            
        let isAccessible (grid: Grid) (roll: PaperRollLocation) =
            getNeighbouringPaperRolls grid roll
                |> List.length < maxAdjacentPaperRolls 
            
        let removeAllAccessibleRolls (grid: Grid) =
            let rec removeAllAccessibleInner removedSoFar =
                let rollsToRemove =
                    grid
                    |> getAllPaperRolls
                    |> List.filter (isAccessible grid)
                
                match rollsToRemove with
                | [] -> removedSoFar
                | _ -> rollsToRemove
                       |> List.iter (fun (row, col) -> grid[row, col] <- Empty)
                       
                       removeAllAccessibleInner (removedSoFar + rollsToRemove.Length)
            
            removeAllAccessibleInner 0
                        
        member _.SolveFirstPart() =
            let grid = readFloorGrid()

            let reachableRolls =
                grid
                |> getAllPaperRolls
                |> List.filter (isAccessible grid)
                |> List.length
                
            writeLine $"Reachable paper rolls: {reachableRolls}"
        
        member _.SolveSecondPart() : unit =
            let grid = readFloorGrid()
            let removed = removeAllAccessibleRolls grid
            
            writeLine $"Removed paper rolls: {removed}"

    type ChallengeSolution04(console: IConsole, reader: ISolutionReader<ChallengeSolution04>) =
        inherit ChallengeSolution<ChallengeSolution04>(console, reader)

        let solution = Solution(reader.ReadLines >> List.ofArray, console.WriteLine)

        override this.SolveFirstPart() = solution.SolveFirstPart()
        override this.SolveSecondPart() = solution.SolveSecondPart()
